pkg/executors/executor_test.go

```diff
@@ -140,3 +144,58 @@ func TestSortValidationErrors(t *testing.T) {
 		})
 	}
 }
+
+// TestExecute_contextCancellation tests that scripts are closed when the
+// context is cancelled.
+func TestExecute_contextCancellation(t *testing.T) {
+	imageName := fmt.Sprintf("shuttle-test-execute-cancellation-%d", time.Now().UnixNano())
+	t.Logf("Starting image %s", imageName)
+	projectContext := config.ShuttleProjectContext{
+		Scripts: map[string]config.ShuttlePlanScript{
+			"serve": {
+				Description: "",
+				Actions: []config.ShuttleAction{
+					{
+						Shell: fmt.Sprintf("docker run --rm -i --name %s nginx", imageName),
+					},
+				},
+				Args: []config.ShuttleScriptArgs{},
+			},
+		},
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		// let the container start before stopping it
+		time.Sleep(500 * time.Millisecond)
+		cancel()
+	}()
+
+	err := Execute(ctx, projectContext, "serve", nil, true)
+	assert.EqualError(t, err, context.Canceled.Error())
+
+	// sadly we need to give the docker some time before "docker ps" shows the
+	// containers
+	time.Sleep(500 * time.Millisecond)
+	images := runningDockerImages(t, imageName)
+	assert.Len(t, images, 0, "expected no images to be running")
+}
+
+func runningDockerImages(t *testing.T, imageName string) []string {
+	t.Helper()
+	cmd := cmd.NewCmd("docker", "ps", "-a", "--format", "{{ .Names }}")
+	status := <-cmd.Start()
+	t.Logf("docker ps: stderr: %v", status.Stderr)
```

Comment:
Do you want to log the error on every run?

Reply:
This is only for debugging purposes to understand what `docker ps` outputs. Usually this is just an empty slice :)

Reply:
👍🏼  Just wanted to make sure it was not left behind in the test.

Reply:
I think it provides value in that if the test fails sometime in the future we have the output at hand instead of having to add it again. Note that by default `go test` only shows test output if `-v` is set or if a test fails :)